src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,185 @@
+            co_await _lock.with([k, next] {
+                // Try connect as a safety check
+                return next->connect().handle_exception(
+                  [k](std::exception_ptr ex) {
+                      // The failed connection is dropped
+                      vlog(
+                        plog.debug,
+                        "Connection for {} already formed. Dropping dup: {}",
+                        k,
+                        ex);
+                  });
+            });
+
+            inner_hash.modify(
+              it_hash, [next](timestamped_user& item) { item.client = next; });
+
+            // Add a prefix to the evicted key so debugging logs
+            // is easier
+            _evicted_items.emplace_back(
+              k + random_generators::gen_alphanum_string(8), old);
+        } else {
+            vlog(plog.debug, "Reuse client for user {}", k);
+        }
+
+        client = it_hash->client;
+
+        // Convert hash iterator to list iterator
+        auto it_list = _cache.project<underlying_list>(it_hash);
+
+        // Update the timestamp and relocate the element
+        // to the front
+        inner_list.modify(it_list, [](timestamped_user& item) {
+            item.last_used = timestamped_user::clock::now();
+        });
+        inner_list.relocate(inner_list.begin(), it_list);
+    }
+
+    co_return client;
+}
+
+namespace {
+template<typename List_Type>
+ss::future<> clean_clients_from_list(
+  List_Type& list, mutex& lock, std::chrono::milliseconds keep_alive) {
+    auto has_expired = timestamped_user::clock::now();
+    auto condition = [keep_alive, has_expired](const timestamped_user& item) {
+        return has_expired >= (item.last_used + keep_alive);
+    };
+
+    auto first = list.begin();
+    auto last = list.end();
+    while (first != last) {
+        if (condition(*first)) {
+            vlog(plog.debug, "Remove {} from cache", first->key);
+            // Stop client before erase
+            co_await lock.with([&first] {
+                return first->client->stop().handle_exception(
+                  [](std::exception_ptr ex) {
+                      // The stop failed
+                      vlog(
+                        plog.debug,
+                        "Stale client stop already happened {}",
+                        ex);
+                  });
+            });
+            list.erase(first++);
+        } else {
+            ++first;
+        }
+    }
+}
+
+template<typename List_Type>
+ss::future<> stop_clients_until_empty(List_Type& list, mutex& lock) {
+    while (!list.empty()) {
+        co_await lock.with([&list] {
+            return list.back().client->stop().handle_exception(
+              [](std::exception_ptr ex) {
+                  // The stop failed
+                  vlog(plog.debug, "Stop already happened {}", ex);
+              });
+        });
+        list.pop_back();
+    }
+}
+} // namespace
```

Comment:
I'm not sure what the purpose of the lock is.

I think the item should be removed from the list so that it can't be looked up, and then stopped.

I can't convince myself that the iterators shouldn't be reevaluated on every iteration.

I think these can both be implemented over something like:
```c++
template<typename List, typename Pred>
ss::future<> remove_client_if(List& list, Pred pred) {
    while (!list.empty()) {
        if (pred(list.front())) {
            auto item = std::move(list.front());
            list.pop_front();
            vlog(plog.debug, "Remove {} from cache", item.key);
            // Stop client before erase
            co_await item.client->stop().handle_exception(
              [&item](std::exception_ptr ex) {
                  // The stop failed
                  vlog(
                    plog.debug,
                    "Stale client {} stop already happened {}",
                    item.key,
                    ex);
              });
        }
    }
}
```

Reply:
Done.

Reply:
Actually, that's clearly an infinite loop.

This should work better:
```c++
template<typename List, typename Pred>
ss::future<> remove_client_if(List& list, Pred pred) {
    std::list<typename List::value_type> remove;
    auto first = list.begin();
    auto last = list.end();
    while (first != last) {
        if (pred(*first)) {
            remove.push_back(std::move(*first));
            list.erase(first++);
        } else {
            ++first;
        }
    }
    for (auto& item : remove) {
        co_await item.client->stop().handle_exception(
          [&item](std::exception_ptr ex) {
              // The stop failed
              vlog(
                plog.debug,
                "Stale client {} stop already happened {}",
                item.key,
                ex);
          });
    }
}
```

Reply:
Is the remove list needed? We could just stop the client before calling list.erase

Reply:
But then there's a scheduling point, during which the iterators can be invalidated. That's why I've made it two-phase.

Reply:
Done.